linkerd/core/src/main/scala/io/buoyant/linkerd/RequestEvaluator.scala

```diff
@@ -0,0 +1,168 @@
+
+  private[this] val RequestTracerMaxDepthHeader = "l5d-max-depth"
+
+  private[this] def printEvaluatedRequest(
+    prependText: String,
+    routerLabel: String,
+    serviceName: String,
+    clientName: String,
+    selectedAddress: String,
+    addresses: Option[Set[String]],
+    dtabResolution: List[String]
+  ) = {
+    Seq(
+      prependText,
+      s"""
+         |--- Router: $routerLabel ---
+         |service name: $serviceName
+         |client name: $clientName
+         |selected address: $selectedAddress
+         |addresses: [${addresses.getOrElse(Set.empty).mkString(", ")}]
+         |dtab resolution:
+         |${dtabResolution.map("  " + _).mkString("\n")}
+    """.stripMargin
+    ).mkString("\n").trim
+  }
+
+  private[this] def formatDTree(
+    dTree: DelegateTree[Name.Bound],
+    tree: List[String],
+    searchPath: Path
+  ): List[String] = {
+    dTree match {
+      case Leaf(path, _, bound) if bound.id == searchPath =>
+        tree :+ path.show
+      case Transformation(_, _, value, remainingTree) =>
+        formatDTree(remainingTree, tree :+ value.path.show, searchPath)
+      case Delegate(path, _, remainingTree) =>
+        formatDTree(remainingTree, tree :+ path.show, searchPath)
+      case Union(path, _, remainingWeightedTrees@_*) =>
+        remainingWeightedTrees.map { wd =>
+          formatDTree(wd.tree, tree :+ s"${wd.weight} * ${path.show}", searchPath)
+        }.filter(!_.isEmpty).head
+      case Alt(path, _, remainingTrees@_*) =>
+        remainingTrees.map { d =>
+          formatDTree(d, tree :+ path.show, searchPath)
+        }.filter(!_.isEmpty).head
+      case _ => List.empty
+    }
+  }
+
+  private[this] def getRouterCtx(prevResp: Response) = {
+
+    val serviceName = DstPathCtx.current match {
+      case Some(dstPath) => dstPath.path
+      case None => Path.empty
+    }
+
+    val selectedEndpoint = endpoint.addr match {
+      case inetAddr: Address.Inet => inetAddr.addr.toString.stripPrefix("/")
+      case _ => ""
+    }
+
+    val clientPath = DstBoundCtx.current match {
+      case None => Path.empty
+      case Some(addrSet) => addrSet.name.id match {
+        case p: Path => p
+        case _ => Path.empty
+      }
+    }
+
+    val lbSet = DstBoundCtx.current match {
+      case None => Future.value(Addr.Neg)
+      case Some(addrSet) => addrSet.name.addr.changes.toFuture
+    }
+
+    val addresses = lbSet.map {
+      case Addr.Bound(a, _) => Some(
+        a.map {
+          case inetAddr: Address.Inet => inetAddr.addr.toString.stripPrefix("/")
+          case _ => ""
+        }
+      )
+      case _ => None
+    }
+
+    val dtreeF = namer.interpreter match {
+      case delegator: Delegator => delegator.delegate(dtab.dtab(), serviceName)
+        .map(Some(_))
+      case _ => Future.None
+    }
+
+    dtreeF.joinWith(addresses) {
+      case (dTree: Option[DelegateTree[Name.Bound]], addrSet: Option[Set[String]]) =>
+        val resp = Response()
```

Comment:
consider just appending to prevResp.contentString instead of creating a new Response